### PR TITLE
Allow single line braces for empty function body

### LINKED ISSUE #5475: [BracesFixer] constructor property promotion: allow braces on same line
## Feature request

I want to write constructors like this:

```php
class Foo
{
    public construct(
        private int $bar;
    ) {}
}
```

But the BracesFixer moves the closing `}` to the next line:

```php
class Foo
{
    public construct(
        private int $bar;
    ) {
    }
}
```

### PR DESCRIPTION
**Pull request by tuqqu**
Closes #5475.

This option allows placing braces on a single line for empty function bodies:

```php
class X 
{
     public function __construct(
              public int $x,
              //...
     ) {}
}
```

Since the introduction of property promotion this kind of constructors are quite widespread. This option also enables this behaviour for other empty functions as well, e.g. 
```php
public function x(): void {}
```

### CODE REVIEW

**Comment by julienfalque** [comment]
As [proposed in #5475](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5475#issuecomment-834302137), it might be better to deprecate the `allow_single_line_anonymous_class_with_empty_body` option to replace it with a new one that would apply to any braces instead of some specific cases. For example, we could go with `allow_single_line_empty_body`. What do you think?

**Comment by tuqqu** [comment]
@julienfalque I think it's a good suggestion. I adjusted the PR. Let me know if there is anything that still needs to be fixed.

**Comment by tuqqu** [comment]
@julienfalque Not sure what to do with tests [complaining](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5757/checks?check_run_id=2728969774#step:11:429) about self-deprecations. I can't remove them, since the option being deprecated had a default value, therefore must still be present (I guess). Any ideas?

**Comment by julienfalque** [comment]
@tuqqu You probably need to update rulesets to stop using the deprecated option. Tests that still use it purposely should be tagged with `@group legacy` and may use `$this->expectDeprecation()`.

File: doc/ruleSets/PSR12.rst [hunk 1]
```diff
@@ -11,7 +11,7 @@ Rules
 - `blank_line_after_opening_tag <./../rules/php_tag/blank_line_after_opening_tag.rst>`_
 - `braces <./../rules/basic/braces.rst>`_
   config:
-  ``['allow_single_line_anonymous_class_with_empty_body' => true]``
+  ``['allow_single_line_empty_body' => true]``
```
Comment (mvorisek) [inline doc/ruleSets/PSR12.rst:14]: empty bodies should be not a default in general, it decreases readability (for ex. on methods) and complicates diffing

Comment (tuqqu) [reply doc/ruleSets/PSR12.rst:14]: Not really. It does not force the rule, it just allows inlining them on purpose.

If we want to update rulesets to not use the newly deprecated option we are left with:
a) removing just the deprecated option, which is a BC break for sure.
b) replacing the deprecated option with the new one. That's not a BC break, since old code will continue to work as it worked.

Comment (julienfalque) [reply doc/ruleSets/PSR12.rst:14]: This option does not force empty bodies to be single line, it only *allows* that. I don't see that as a BC break.

Though this makes me think: since this codestyle is becoming more common, maybe we should propose an option to actually force that. So instead of `allow_single_line_empty_body => true|false`, maybe we could go with e.g. `empty_body => ignore|multi_line|single_line`?

File: tests/Fixer/Basic/BracesFixerTest.php [hunk 2]
```diff
@@ -4696,21 +4696,21 @@ public function sth(): string
 $foo = new class () extends \Exception {};
 ',
                 null,
-                ['allow_single_line_anonymous_class_with_empty_body' => true],
```
Comment (julienfalque) [inline tests/Fixer/Basic/BracesFixerTest.php:4699]: Those tests should not be changed: the deprecated option should still work as expected until we actually remove it.

File: tests/Fixer/Basic/BracesFixerTest.php [hunk 3]
```diff
@@ -4888,6 +4888,150 @@ public function provideFixWithAllowOnelineLambdaCases()
...
+    try {
+        0;
+    } catch (\Throwable $t) {}',
+                '<?php
+    try { 0; } catch (\Throwable $t) {}',
+            ],
+            [
+                '<?php
+    try {} catch (\Throwable $t) {}',
+            ],
```
Comment (julienfalque) [inline tests/Fixer/Basic/BracesFixerTest.php:5031]: Please add a test with a multi line empty body that is left unchanged:
```
function foo() {
}
```

File: tests/Fixtures/Integration/misc/PHP7_0.test [hunk 4]
```diff
@@ -75,14 +75,10 @@ class Foo implements FooInterface
     }
 
     // scalar typehinting
-    public function bar1(int $a): string
-    {
-    }
+    public function bar1(int $a): string {}
```
Comment (julienfalque) [inline tests/Fixtures/Integration/misc/PHP7_0.test:78]: Why do integration tests need to be updated? Doesn't the new option ignore multi line empty bodies?

File: tests/Fixtures/Integration/set/@Symfony.test-in.php [hunk 5]
```diff
@@ -65,7 +65,7 @@ private function transformText($dummy, array $options = array(),$data=null)
             return ucwords($dummy);
         }
 
-        throw new \RuntimeException(sprintf('Unrecognized dummy option "%s"', $dummy));
+        throw new \RuntimeException(sprintf('Unrecognized dummy option "%s".', $dummy));
```
Comment (julienfalque) [inline tests/Fixtures/Integration/set/@Symfony.test-in.php:68]: Is this change actually related to the new option?

File: doc/ruleSets/PSR12.rst (hunk 1 above)
Comment (mvorisek) [reply doc/ruleSets/PSR12.rst:14]: 👍 , but for anonymous language constructs (anonymous function, class) there should be a separate configuration.

And for anonymous `single_line` should be allowed, for non-anonymous (class, ifaces, function, methods) `multi_line` should be enforced.

Comment (julienfalque) [reply doc/ruleSets/PSR12.rst:14]: Why the distinction between anonymous and non-anonymous?

Comment (mvorisek) [reply doc/ruleSets/PSR12.rst:14]: Because forcing `{}` to be on a single line for classes, methods etc. reduces readablity and complicates diffing.

Comment (julienfalque) [reply doc/ruleSets/PSR12.rst:14]: This applies to anonymous functions and classes.

File: tests/Fixer/Basic/BracesFixerTest.php (hunk 2 above)
Comment (tuqqu) [reply tests/Fixer/Basic/BracesFixerTest.php]: ok, I changed it back. But this triggers the `Remaining self deprecation notices` message and fails the tests. I added the `SYMFONY_DEPRECATIONS_HELPER` to `phpunit.xml` help with that. 

Please, let me know if there should be another solution to this.

File: tests/Fixtures/Integration/misc/PHP7_0.test (hunk 4 above)
Comment (tuqqu) [reply tests/Fixtures/Integration/misc/PHP7_0.test:78]: The new option indeed does ignore it. 

But those tests in fact have inlined `foo() {}` in the `INPUT` section, so I adjusted the `EXPECT` section accordingly.

File: tests/Fixtures/Integration/set/@Symfony.test-in.php (hunk 5 above)
Comment (tuqqu) [reply tests/Fixtures/Integration/set/@Symfony.test-in.php]: It is not, but fabbot.io starts [nudging](https://fabbot.io/report/FriendsOfPHP/PHP-CS-Fixer/5757/34120b69992a615c906195243a6fda1143354343) about those files once I touch them. 

What should I do about it?

File: tests/Fixer/Basic/BracesFixerTest.php (hunk 2 above)
Comment (julienfalque) [reply tests/Fixer/Basic/BracesFixerTest.php]: There's a better solution: the test cases using the deprecated option should be moved to a dedicated test method with a `@legacy` annotation and with a `$this->expectDeprecation()` call. You can find examples in other tests in the project.

File: tests/Fixtures/Integration/set/@Symfony.test-in.php (hunk 5 above)
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@Symfony.test-in.php]: It happens sometimes, I'd say we can ignore that.

**Comment by tuqqu** [comment]
@julienfalque thanks for the review. I made changes to the code. I tested on my machine and all tests are passing now.

**Comment by tuqqu** [comment]
@julienfalque I'm not sure if this [fail](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5757/checks?check_run_id=2780875027) relates to this PR in any way. What should we do?

**Comment by julienfalque** [comment]
@tuqqu Indeed, [the issue also happens on `master` branch](https://github.com/FriendsOfPHP/PHP-CS-Fixer/runs/2747753773). We should fix that so you can rebase and have a green build :)

**Comment by julienfalque** [comment]
@tuqqu Build fixed on `master`, please rebase.

**Comment by tuqqu** [comment]
@julienfalque, the tests are passing now, the only issue is with fabbot.io that you suggested to ignore.

is there anything else that ought to be done with this PR?

**Comment by julienfalque** [comment]
We didn't make a decision about https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5757#discussion_r646499681.

**Comment by tuqqu** [comment]
@julienfalque Hello. Is there any estimation of how long it will take to make a decision?

**Comment by tuqqu** [comment]
@keradus Maybe you could help with that question? Just want to make sure this PR is not forgotten

**Comment by GrahamCampbell** [comment]
👎 on the options here. We should instead add a separate option specifically for saying if this should be allowed for constructors only, since they are special methods that support property promotion. Maybe we should be even stricter, and only allow this if property promotion is actually present!

**Comment by ralphjsmit** [comment]
Any idea when this will become an option in PHP CS Fixer?

**Comment by gnutix** [comment]
Very eager to see this one coming out too !

**Comment by ralphjsmit** [comment]
@julienfalque @SpacePossum Could you shed some light on the release date/timeline of this rule? Thanks!🙌

**Comment by winkbrace** [comment]
My team would much appreciate this option too. :)

And I know this is out of scope, but FYI another use case for empty body I sometimes have is 

```php
class CustomException extends \Exception 
{}
```

**Comment by SpacePossum** [comment]
just a FYI as it is `braces related` I won't be the one reviewing it, there are however two unresolved comments from reviewers here.

**Comment by julienfalque** [comment]
@tuqqu The handling of braces position has been extracted from `braces` into `curly_braces_position`, can you rebase and move your changes to this new rule?

**Comment by tuqqu** [comment]
@julienfalque it's been a while, I deleted the branch since. Whoever wants to take on the issue is welcome

**Comment by Crell** [comment]
Just a note that the next version of PER-CS is going to encourage (but not require) the style shown here.

**Comment by Wirone** [comment]
@tuqqu would you mind to rebase your branch and refresh this PR? I think it's worth finishing, since it's second most liked issue on the backlog 🙂.

**Comment by nickygerritsen** [comment]
I tried rebasing this myself but I got quite stuck with the logic in the new curly braceposition fixer. I tried setting the `$allowSingleLineIfEmpty` by default to the new config instead of `false`, but that makes it do this:

```
function test() {}
```
becomes
```
function test()
{}
```

I do hope we can get this fixed, since I also think the empty body case is a good one.

**Comment by Wirone** [comment]
@tuqqu's fork was removed, but it could be picked up, **BUT** I believe this should be closed since it provides changes for `braces` fixer which was deprecated and is not supported anymore. We need new implementation based on smaller, existing fixers or new ones. @keradus @kubawerlos what do you think?

**Comment by Wirone** [comment]
Closing this one, we'll continue in #6933 🙂
